### ISSUE TITLE
Machine ID: Support delegated joining on GitLab

### ISSUE THREAD

**Issue by strideynet** [gravitational/teleport#17059]
Reference GitHub implementation at https://github.com/gravitational/teleport/pull/16938

https://docs.gitlab.com/ee/ci/cloud_services/

https://docs.gitlab.com/ee/ci/variables/index.html#pass-an-environment-variable-to-another-job

### Information available in GitLab tokens

CI_JOB_JWT:
```json
{
  "alg": "RS256",
  "kid": "4i3sFE7sxqNPOT7FdvcGA1ZVGGI_r-tsDXnEuYT4ZqE",
  "typ": "JWT"
},
{
  "namespace_id": "1008548",
  "namespace_path": "strideynet",
  "project_id": "39980849",
  "project_path": "strideynet/test-gitlab-jwt",
  "user_id": "842508",
  "user_login": "strideynet",
  "user_email": "redacted",
  "pipeline_id": "659183071",
  "pipeline_source": "push",
  "job_id": "3131574536",
  "ref": "main",
  "ref_type": "branch",
  "ref_protected": "true",
  "jti": "03362687-b43c-404a-b3bd-0a57e2f7b512",
  "iss": "gitlab.com",
  "iat": 1664985077,
  "nbf": 1664985072,
  "exp": 1664988677,
  "sub": "job_3131574536"
}
```

CI_JOB_JWT_V2
```json
{
  "alg": "RS256",
  "kid": "4i3sFE7sxqNPOT7FdvcGA1ZVGGI_r-tsDXnEuYT4ZqE",
  "typ": "JWT"
},
{
  "namespace_id": "1008548",
  "namespace_path": "strideynet",
  "project_id": "39980849",
  "project_path": "strideynet/test-gitlab-jwt",
  "user_id": "842508",
  "user_login": "strideynet",
  "user_email": "redacted",
  "pipeline_id": "659183071",
  "pipeline_source": "push",
  "job_id": "3131574536",
  "ref": "main",
  "ref_type": "branch",
  "ref_protected": "true",
  "jti": "f5e60e65-c5c8-4a27-a4d9-e689d580d084",
  "iss": "https://gitlab.com",
  "iat": 1664985077,
  "nbf": 1664985072,
  "exp": 1664988677,
  "sub": "project_path:strideynet/test-gitlab-jwt:ref_type:branch:ref:main",
  "aud": "https://gitlab.com"
}
```

**Comment by mdaniel** [comment]
Even if the full-blown CEL implementation doesn't make it in, please allow either substring matches, globbing, or regex, because within GitLab the hierarchy is a lot more complex than GitHub, and thus just trusting `repository_owner:` or one-by-one `repository:` items will be painful in most GL setups

**Comment by strideynet** [comment]
> Even if the full-blown CEL implementation doesn't make it in, please allow either substring matches, globbing, or regex, because within GitLab the hierarchy is a lot more complex than GitHub, and thus just trusting `repository_owner:` or one-by-one `repository:` items will be painful in most GL setups

Indeed, it looks as if really we need to let users glob/regex on `sub` since GitLab doesn't seem to offer more broken down claims like GitHub do. Thanks for the feedback on this.

**Comment by rosskusler** [comment]
I too would love a way for Gitlab-ci jobs running in containerized/ephemeral environments (with no access to a long-running tbot daemon) to be able to leverage machine-id. If Teleport could be configured to trust the Gitlab signed JWT that would be really useful, I think.

**Comment by strideynet** [comment]
https://gitlab.com/gitlab-org/gitlab/-/merge_requests/72555

**Comment by strideynet** [comment]
https://docs.gitlab.com/ee/ci/yaml/index.html#id_tokens

```yaml
job_needing_oidc_auth:
  id_tokens:
    OIDC_TOKEN:
      aud: https://oidc.provider.com
  script:
    - echo $OIDC_TOKEN
```

Supported from 15.7 to allow audience override. We should enforce this.

**Comment by strideynet** [comment]
Note: we should ensure the minimum requirement of GitLab 15.7 is communicated.

**Comment by strideynet** [comment]
Released in 12.2
